libafl_cc/src/tables-pass.cc

```diff
@@ -0,0 +1,229 @@
+              }
+            });
+  #endif
+          }};
+}
+#else
+
+char TamingParsingTables::ID = 0;
+#endif
+
+#ifdef USE_NEW_PM
+PreservedAnalyses TamingParsingTables::run(Module &M, ModuleAnalysisManager &MAM) {
+#else
+bool TamingParsingTables::runOnModule(Module &M) {
+#endif
+
+  LLVMContext &C = M.getContext();
+
+  IntegerType *Int64Ty = IntegerType::getInt64Ty(C);
+  IntegerType *Int32Ty = IntegerType::getInt32Ty(C);
+  IntegerType *Int16Ty = IntegerType::getInt16Ty(C);
+  IntegerType *Int8Ty = IntegerType::getInt8Ty(C);
+  IntegerType *Int1Ty = IntegerType::getInt1Ty(C);
+  Type *VoidTy = Type::getVoidTy(C);
+
+  FunctionCallee LogFunc = M.getOrInsertFunction("__libafl_tables_transition", VoidTy, Int32Ty, Int32Ty);
+
+#ifdef USE_NEW_PM
+  auto PA = PreservedAnalyses::all();
+#endif
+
+  /* Instrument all the things! */
+
+  for (auto &F : M) {
+    int has_calls = 0;
+
+    // if (!isInInstrumentList(&F)) { continue; }
+
+    if (F.size() < function_minimum_size) { continue; }
+
+    std::unordered_set<LoadInst *> loads;
+    std::unordered_set<Value *>    geps;
+
+    for (auto &BB : F) {
+      for (auto &I : BB) {
+        GetElementPtrInst *GEP;
+        StoreInst         *ST;
+        LoadInst          *LI;
+        if ((LI = dyn_cast<LoadInst>(&I))) {
+          loads.insert(LI);
+        } else if ((GEP = dyn_cast<GetElementPtrInst>(&I))) {
+          if (!GEP->hasIndices() || GEP->hasAllConstantIndices()) continue;
+          // TODO handle multiple idxs
+          Value *IDX = *GEP->idx_begin();
+          IDX = recurseCast(IDX);
+          
+          if ((LI = dyn_cast<LoadInst>(IDX)) && loads.find(LI) != loads.end()) {
+            geps.insert(GEP);
+          }
+        } else if ((ST = dyn_cast<StoreInst>(&I))) {
+          Value    *PTR = ST->getPointerOperand();  // must be in a prev load
+          Value    *VAL = recurseCast(ST->getValueOperand());
+          LoadInst *GL = nullptr;
+          Value    *V = nullptr;
+          
+          if ((GL = dyn_cast<LoadInst>(VAL))) {
+            V = GL->getPointerOperand();
+            if (V == nullptr || geps.find(V) == geps.end()) { continue; }
+          } else {
+            continue;
+          }
+
+          // the value comes from a load in which the ptr is obtained with a gep
+          GEP = (GetElementPtrInst *)V;
+          Value *IDX = *GEP->idx_begin();
+          IDX = recurseCast(IDX);
+          LI = dyn_cast<LoadInst>(IDX);
+
+          if (LI == nullptr) continue;
+          if (recurseCast(LI->getPointerOperand()) !=
+              recurseCast(ST->getPointerOperand()))
+            continue;
+
+          std::string location = std::string("UNKNOWN");
+          if (DILocation *Loc = GEP->getDebugLoc().get()) {
+            location = std::string(Loc->getFilename().data()) +
+                       std::string(":") + std::to_string(Loc->getLine());
+          }
+
+          errs() << "FOUND " << location << "\n\t" << *LI << "\n\t" << *GEP
```

Comment:
i think stderr will make fuzzebench fuzzer fail to build

Reply:
i know man, it will be under the debug option but I'm still testing it atm